src/tools/compiletest/src/header/auxiliary.rs

```diff
@@ -63,3 +67,124 @@ fn parse_aux_crate(r: String) -> (String, String) {
         parts.next().expect("missing aux-crate value (e.g. log=log.rs)").to_string(),
     )
 }
+
+/// Return an error if the given directory has cyclic aux.
+pub(crate) fn check_cycles(config: &Config, dir: &Path) -> io::Result<()> {
+    let mut filenames = vec![];
+    let mut auxiliaries = HashMap::new();
+
+    build_graph(config, dir, dir, &mut filenames, &mut auxiliaries)?;
+
+    has_cycle(&filenames, &auxiliaries)
+}
+
+fn build_graph(
+    config: &Config,
+    dir: &Path,
+    base_dir: &Path,
+    filenames: &mut Vec<String>,
+    auxiliaries: &mut HashMap<String, Vec<String>>,
+) -> io::Result<()> {
```

Comment:
Problem: so I thought about this a bit. `build_graph` during test collection that does directory traversal and file reading and prop parsing is *quite* expensive (especially on slower filesystems like native Windows), and I don't think this is necessarily the right place to do the check.

I would've expected that compiletest does something like:

1. First, it collects the tests based on the path filter (or was it test name filter) that was passed from bootstrap.
2. Then, compiletest will do some "early props" collection and handling. At this point is when I believe compiletest establishes the basic test conditions and relationships between main test files and their auxiliaries (i.e. if a test will be run or ignored, if it has auxiliaries and such).
3. Then, compiletest will further collect and handle "late props".

I would've expected that aux cycle detection happens *after* early prop collection after we establish the relationship between test files and auxiliaries, and not during test discovery and collection.

Alternatively, we could also just declare that we're not going to handle cyclic auxiliaries and "just don't do that", and doing cyclic auxiliaries is considered PEBKAC as "compiletest did exactly what the user instructed it to do".

cc @oli-obk: how does `ui_test` handle infinitely-recursive auxiliaries?

Reply:
It doesn't: https://github.com/oli-obk/ui_test/issues/123

But it should be fixable now that I have a build manager that handles all of these in one place.

Reply:
@jieyouxu 
Thanks for your review.

> 2. Then, compiletest will do some "early props" collection and handling. At this point is when I believe compiletest establishes the basic test conditions and relationships between main test files and their auxiliaries (i.e. if a test will be run or ignored, if it has auxiliaries and such).

Now I understood that cycle detection should be run around [here](https://github.com/rust-lang/rust/blob/master/src/tools/compiletest/src/lib.rs#L813-L825), is that right?

FYI: EarlyProps gathers auxiliaries only which are referred from the test file. For example from my manual test files above, `cycle1.rs` is included in the early_props for `aux-cyclic.rs` but `cycle2.rs` is not because only cycle1.rs has a dependency to cycle2.rs. So reading files to make a dependency graph is inevitable, but it should be lighter than reading files by entire directory traversal since we can check the cycle only against test files which contains auxs.

Reply:
Good question, I'll have to revisit this.